### PR TITLE
Add new openpolicyagent connector (V2)

### LINKED ISSUE #702: Add new openpolicyagent connector (V2)
Add new openpolicyagent connector in connectors/ (just its go code and Dockerfile)

### PR DESCRIPTION
**Pull request by rohithdv**
This PR fixes #702.
This PR has the following changes: 
(1) Adds a `lib` folder in `connectors/open_policy_agent` which has functionality to contact catalog connector and opa to retrieve policy decisions. This folder is based on the `connectors/opa/lib` folder currently used in the GRPC based opa connector. 

(2) Updates the `connectors/open_policy_agent/testutil `to be compatible with openapi taxonomy related changes. 

(3) changes the codebase in `connectors/open_policy_agent/lib` to be compatible with openapi taxonomy related changes. 

(4) Updates the `api_default.go` in `connectors/open_policy_agent/go` to have functionality of contacting OPA once the Openapi based HTTP Request arrives in the gin based http server (which is the V2 opa connector) 

(5) `GetPoliciesDecisions` function in `api_default.go` now returns back an HTTP response which contains the policy manager response object (which is compatible with the openapi based taxonomy). 

(6) Added `charts/fybrik/templates/openpolicyagent-connector` folder which has the helm based files to start V2 OPA connector (Openapi based HTTP server ) during the install of fybrik (i.e. when the command  `helm install fybrik charts/fybrik --set global.tag=latest -n fybrik-system --wait`) is run. 

**PS:** Currently during fybrik installation, we are deploying V1 and V2 versions of OPA Connector to check compatibility with GRPC and OpenAPI protocols. We will remove the V1 opa connector in a subsequent PR once the end to end flow of Openapi based taxonomy is fully tested. 

To summarize at a high level, currently the fybrik deployment looks like below:

`pod/openpolicyagent-connector-84bfc6f495-2z5n6 `refers to V2 OPA connector (Openapi HTTP based)
`pod/opa-connector-674d8d489d-rmwqg`  refers to V1 OPA Connector (GRPC based)
`pod/manager-client-sn7bz`  is a test client which mocks the manager functionality related to openapi taxonomy changes. The `manager-client`  is a temporary installation only for testing purposes and will be removed in a cleanup PR. 

image

### CODE REVIEW

**Review by ritwikc — APPROVED** [review]

File: connectors/open_policy_agent/go/api_default.go [hunk 1]
```diff
@@ -13,18 +13,109 @@
...
+
+func displayReq(input2 openapiclientmodels.PolicyManagerRequest) error {
+	bytes, err := input2.MarshalJSON()
+	if err != nil {
+		log.Println("err: ", err)
+		return nil
+	}
+	log.Println("marshalled request:", string(bytes))
+	return nil
+}
```
Comment (roee88) [inline connectors/open_policy_agent/go/api_default.go:60]: I believe we implemented such debug functions several times already. I'm not even sure if it's needed and printing with %+v would be enough but in any case a function like that should move to `pkg/` and be reused across the project. Look if we have an implementation already in manage (cc @shlomitk1). 

I'm personally not a fan of these functions since they require marshalling just for debugging. I would prefer to use a gin middleware instead for logging both request and response and add a flag to either add the middleware or not.

**Comment by ritwikc** [comment]
Just a few questions/comments I am putting here.:
1. In opa_reader.go do you want to print the tags and the values ? I know those are for debugging. But are those needed in the merged code.
2. In  test/services/pilot/mock-openapi-policymanager/manager-client.go I see the <SECRET-NAME> and <NAMESPACE> are hard coded.
3. In charts/fybrik/values.yaml on line 475 you are referring to rohithdv. Is that what it is supposed to be ?

File: connectors/open_policy_agent/go/api_default.go (hunk 1 above)
Comment (rohithdv) [reply connectors/open_policy_agent/go/api_default.go]: @roee88 is it ok to use this library [render](https://github.com/luci/go-render) as suggested [here](https://stackoverflow.com/questions/43259971/golang-how-to-print-struct-value-with-pointer) ? Currently I have used this library in the opa connector as shown [here](https://github.com/rohithdv/fybrik/blob/186fcd6c2bc818754c825f777e9fe0843c5eaf27/connectors/open_policy_agent/go/api_default.go#L57).

or do you suggest some other way for printing structs with pointers ? Please advise. Thanks.

File: charts/fybrik/templates/openpolicyagent-connector/openpolicyagent-connector-cm.yaml [hunk 2]
```diff
@@ -0,0 +1,13 @@
+{{- $autoFlag := and .Values.coordinator.enabled (eq .Values.coordinator.policyManager "opa") }}
```
[the remark this answers is not in the document]
Comment (rohithdv) [reply charts/fybrik/templates/openpolicyagent-connector/openpolicyagent-connector-cm.yaml]: Fixed. Please check.

File: charts/fybrik/templates/openpolicyagent-connector/openpolicyagent-connector-deployment.yaml [hunk 3]
```diff
@@ -0,0 +1,63 @@
...
+        {{- toYaml .Values.openPolicyAgentConnector.podSecurityContext | nindent 8 }}
+      containers:
+        - name: openpolicyagent-connector
+          securityContext:
+            {{- toYaml .Values.openPolicyAgentConnector.securityContext | nindent 12 }}
+          image: {{ include "fybrik.image" ( tuple $ .Values.openPolicyAgentConnector ) }}
+          imagePullPolicy: {{ .Values.openPolicyAgentConnector.imagePullPolicy | default .Values.global.imagePullPolicy }}
+          ports:
+            - name: http
+              containerPort: 8081
```
[the remark this answers is not in the document]
Comment (rohithdv) [reply charts/fybrik/templates/openpolicyagent-connector/openpolicyagent-connector-deployment.yaml]: Fixed. Please check.

File: connectors/open_policy_agent/Dockerfile [hunk 4]
```diff
@@ -0,0 +1,7 @@
+FROM gcr.io/distroless/static:nonroot
+
+COPY openpolicyagent-connector /
+
+EXPOSE 50082
```
[the remark this answers is not in the document]
Comment (rohithdv) [reply connectors/open_policy_agent/Dockerfile]: Fixed. Please check.

File: connectors/open_policy_agent/go/api_default.go [hunk 5]
```diff
@@ -13,18 +13,109 @@
...
 )
 
+func getEnv(key string) string {
+	value, exists := os.LookupEnv(key)
+	if !exists {
+		log.Fatalf("Env Variable %v not defined", key)
+	}
+	log.Printf("Env. variable extracted: %s - %s\n", key, value)
+	return value
+}
```
[the remark this answers is not in the document]
Comment (rohithdv) [reply connectors/open_policy_agent/go/api_default.go]: I have changed the implementation. Please review. Thanks.

File: connectors/open_policy_agent/go/api_default.go [hunk 6]
```diff
@@ -13,18 +13,109 @@
...
+	log.Println(inputString)
+	var input openapiclientmodels.PolicyManagerRequest
+	err := json.Unmarshal([]byte(inputString), &input)
+	if err != nil {
+		return nil, err
+	}
+	log.Printf("input: %v", input)
+
+	return &input, nil
+}
```
[the remark this answers is not in the document]
Comment (rohithdv) [reply connectors/open_policy_agent/go/api_default.go]: Fixed. Please check.

File: connectors/open_policy_agent/go/api_default.go [hunk 7]
```diff
@@ -13,18 +13,109 @@
...
+	}
+
+	jsonOutput, err := json.MarshalIndent(eval, "", "\t")
+	if err != nil {
+		return openapiclientmodels.PolicyManagerResponse{}, fmt.Errorf("error during MarshalIndent of OPA decisions: %v", err)
+	}
+	log.Println("contactOPA: Received evaluation after execution of GetOPADecisions : " + string(jsonOutput))
+
+	return eval, nil
+}
```
[the remark this answers is not in the document]
Comment (rohithdv) [reply connectors/open_policy_agent/go/api_default.go]: Most of the statements in this function (not related to catalog connector and opa server) are now removed. I have kept the function still with the necessary modification. Please suggest if I need to alter anything else. Thanks.

**Comment by rohithdv** [comment]
> 2\. see the  and  are hard code

Addressed all the comments except (1). I think the log statements are useful till we test end to end flow. Once that is done, we can prune unnecessary logs. Hope this is fine for now. Thanks.

File: connectors/open_policy_agent/go/api_default.go [hunk 8]
```diff
@@ -13,18 +13,54 @@
...
+	"github.com/gdexlab/go-render/render"
 	"github.com/gin-gonic/gin"
 )
 
-// GetPoliciesDecisions - getPoliciesDecisions
-func GetPoliciesDecisions(c *gin.Context) {
+type ConnectorController struct {
+	Policytobeevaluated string
+	Opaserverurl        string
+	Opaclient           *http.Client
```
Comment (roee88) [inline connectors/open_policy_agent/go/api_default.go:28]: Use CamelCase (current names are unreadable)

File: connectors/open_policy_agent/go/api_default.go [hunk 9]
```diff
@@ -13,18 +13,54 @@
...
+	opabl "fybrik.io/fybrik/connectors/open_policy_agent/lib"
+	openapiclientmodels "fybrik.io/fybrik/pkg/taxonomy/model/base"
+	"github.com/gdexlab/go-render/render"
 	"github.com/gin-gonic/gin"
 )
 
-// GetPoliciesDecisions - getPoliciesDecisions
-func GetPoliciesDecisions(c *gin.Context) {
+type ConnectorController struct {
+	Policytobeevaluated string
```
Comment (roee88) [inline connectors/open_policy_agent/go/api_default.go:26]: Why do we need this?

File: connectors/open_policy_agent/go/routers.go [hunk 10]
```diff
@@ -18,59 +18,15 @@ import (
...
-
+	router.POST("/", Index)
+	router.POST("/getPoliciesDecisions", connController.GetPoliciesDecisions)
 	return router
 }
 
 // Index is the index handler.
 func Index(c *gin.Context) {
 	c.String(http.StatusOK, "Hello World!")
 }
```
Comment (roee88) [inline connectors/open_policy_agent/go/routers.go:32]: Remove?

File: connectors/open_policy_agent/go/api_default.go [hunk 11]
```diff
@@ -13,18 +13,54 @@
...
+	Opaserverurl        string
+	Opaclient           *http.Client
+	CatalogReader       *opabl.CatalogReader
+}
+
+func (e *ConnectorController) contactOPA(input openapiclientmodels.PolicyManagerRequest, creds string) (openapiclientmodels.PolicyManagerResponse, error) {
+	opaReader := opabl.NewOpaReader(e.Opaserverurl)
```
Comment (roee88) [inline connectors/open_policy_agent/go/api_default.go:33]: Reuse opa client instead of creating a new one on each call

File: connectors/open_policy_agent/go/api_default.go [hunk 12]
```diff
@@ -13,18 +13,54 @@
...
+
+	eval, err := opaReader.GetOPADecisions(&input, creds, e.CatalogReader, e.Policytobeevaluated, e.Opaclient)
+	if err != nil {
+		log.Println("GetOPADecisions err:", err)
```
Comment (roee88) [inline connectors/open_policy_agent/go/api_default.go:37]: Don't log returned errors. You may use errors.Wrap instead if you need to add information to the error.

File: connectors/open_policy_agent/go/api_default.go (hunk 8 above)
Comment (rohithdv) [reply connectors/open_policy_agent/go/api_default.go]: fixed. Please check. Thanks.

File: connectors/open_policy_agent/go/routers.go (hunk 10 above)
Comment (rohithdv) [reply connectors/open_policy_agent/go/routers.go]: fixed. Please check. Thanks.

File: connectors/open_policy_agent/go/api_default.go (hunk 12 above)
Comment (rohithdv) [reply connectors/open_policy_agent/go/api_default.go]: fixed. Please check. Thanks.

File: connectors/open_policy_agent/go/api_default.go (hunk 11 above)
Comment (rohithdv) [reply connectors/open_policy_agent/go/api_default.go]: fixed. Please check. Thanks.

File: charts/fybrik/values.yaml [hunk 13]
```diff
@@ -59,11 +59,11 @@ coordinator:
 
   # Configures the policy manager system name to be used by the coordinator manager.
   # Accepted values are "opa" or any meaningful name if a third party connector is used.
-  policyManager: "opa"
+  policyManager: "openpolicyagent"
 
   # Overrides the policy manager connector URL.
   # Defaults to `<policyManager>-connector:80`.
-  policyManagerConnectorURL: ""
+  policyManagerConnectorURL: "http://openpolicyagent-connector:80"
```
[the remark this answers is not in the document]
Comment (rohithdv) [reply charts/fybrik/values.yaml]: fixed. Please check. Thanks.

File: connectors/open_policy_agent/go/api_default.go (hunk 9 above)
Comment (rohithdv) [reply connectors/open_policy_agent/go/api_default.go]: made Policytobeevaluated to be a constant now. Thanks.

**Review by roee88 — APPROVED** [review]
